### PR TITLE
Factor out WebView Debugging

### PR DESCRIPTION
**Pull request by david-allison-1**
## Purpose / Description
I want to use this code in the Visual Editor. 

I also lowkey want to make AbstractFlashCardViewer have less code as it's hard to understand.

## Approach
Extract method, and add a call on startup.

## How Has This Been Tested?
Printf performance testing, not great, but didn't want to spend too much effort

```
On
2020-04-10 18:14:14.086 14245-14245/com.ichi2.anki V/AbstractFlashcardViewer: WebView Debugging Start
2020-04-10 18:14:14.089 14245-14245/com.ichi2.anki V/AbstractFlashcardViewer: WebView Debugging End

Off
2020-04-10 18:13:21.455 12541-12541/com.ichi2.anki V/AbstractFlashcardViewer: WebView Debugging Start
2020-04-10 18:13:21.456 12541-12541/com.ichi2.anki V/AbstractFlashcardViewer: WebView Debugging End

On+Startup
2020-04-10 18:10:45.411 10258-10258/com.ichi2.anki V/AbstractFlashcardViewer: WebView Debugging Start
2020-04-10 18:10:45.412 10258-10258/com.ichi2.anki V/AbstractFlashcardViewer: WebView Debugging End

Off+Startup
2020-04-10 18:11:44.948 11566-11566/com.ichi2.anki V/AbstractFlashcardViewer: WebView Debugging Start
2020-04-10 18:11:44.949 11566-11566/com.ichi2.anki V/AbstractFlashcardViewer: WebView Debugging End
```

## Checklist
- [x] You have not changed whitespace unnecessarily (it makes diffs hard to read)
- [x] You have a descriptive commit message with a short title (first line, max 50 chars).
- [x] Your code follows the style of the project (e.g. never omit braces in `if` statements) 
- [x] You have commented your code, particularly in hard-to-understand areas
- [x] You have performed a self-review of your own code

### CODE REVIEW

File: AnkiDroid/src/main/java/com/ichi2/utils/WebViewDebugging.java [hunk 1]
```diff
@@ -0,0 +1,17 @@
+package com.ichi2.utils;
+
+import android.content.SharedPreferences;
+import android.os.Build;
+import android.webkit.WebView;
+
+public class WebViewDebugging {
```
Comment (timrae) [inline AnkiDroid/src/main/java/com/ichi2/utils/WebViewDebugging.java:7]: Codacy Issue found: [The utility class name 'WebViewDebugging' doesn't match '[A-Z][a-zA-Z0-9]+(Utils?|Helper)'](https://app.codacy.com/manual/timrae/Anki-Android/pullRequest?prid=5341529)

**Comment by timrae** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 2
           

Complexity increasing per file
==============================
- AnkiDroid/src/main/java/com/ichi2/utils/WebViewDebugging.java  2
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/timrae/Anki-Android/pullRequest?prid=5341529&bid=17487717)

File: AnkiDroid/src/main/java/com/ichi2/anki/AnkiDroidApp.java [hunk 2]
```diff
@@ -28,23 +28,20 @@
...
+import com.ichi2.anki.analytics.UsageAnalytics;
 import com.ichi2.anki.exception.ManuallyReportedException;
 import com.ichi2.anki.exception.StorageAccessException;
 import com.ichi2.anki.services.BootService;
 import com.ichi2.anki.services.NotificationService;
 import com.ichi2.compat.CompatHelper;
 import com.ichi2.utils.LanguageUtil;
-import com.ichi2.anki.analytics.UsageAnalytics;
 import com.ichi2.utils.Permissions;
+import com.ichi2.utils.WebViewDebugging;
```
Comment (timrae) [inline AnkiDroid/src/main/java/com/ichi2/anki/AnkiDroidApp.java:44]: Codacy Issue found: [Avoid unused imports such as 'com.ichi2.utils.WebViewDebugging'](https://app.codacy.com/manual/timrae/Anki-Android/pullRequest?prid=5341529)

File: AnkiDroid/src/main/java/com/ichi2/anki/AnkiDroidApp.java [hunk 3]
```diff
@@ -227,6 +228,8 @@ public void onCreate() {
             UsageAnalytics.setDryRun(true);
         }
 
+        //WebViewDebugging.initializeDebugging(preferences);
```
Comment (mikehardy) [inline AnkiDroid/src/main/java/com/ichi2/anki/AnkiDroidApp.java:231]: no quibble on the goal of the PR, but adding commented out code 🤔

Comment (david-allison-1) [reply AnkiDroid/src/main/java/com/ichi2/anki/AnkiDroidApp.java:231]: 🤦 Thanks

Comment (david-allison-1) [reply AnkiDroid/src/main/java/com/ichi2/anki/AnkiDroidApp.java:231]: Codacy would have caught that one. Would it be possible for you to email a key to check locally?

File: AnkiDroid/src/main/java/com/ichi2/anki/AnkiDroidApp.java [hunk 4]
```diff
@@ -62,7 +59,11 @@
 import java.util.regex.Matcher;
 import java.util.regex.Pattern;
 
+import androidx.annotation.NonNull;
```
Comment (david-allison-1) [inline AnkiDroid/src/main/java/com/ichi2/anki/AnkiDroidApp.java:62]: Android Studio makes these changes consistently in a few files within the project. I'll revert for now; are you using other settings to keep these in place?

**Comment by timrae** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 1
           

Complexity increasing per file
==============================
- AnkiDroid/src/main/java/com/ichi2/utils/WebViewDebugging.java  2
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/timrae/Anki-Android/pullRequest?prid=5341529&bid=17487717)

File: AnkiDroid/src/main/java/com/ichi2/utils/WebViewDebugging.java (hunk 1 above)
Comment (timrae) [inline AnkiDroid/src/main/java/com/ichi2/utils/WebViewDebugging.java:7]: Codacy Issue found: [The utility class name 'WebViewDebugging' doesn't match '[A-Z][a-zA-Z0-9]+(Utils?|Helper)'](https://app.codacy.com/manual/timrae/Anki-Android/pullRequest?prid=5341529)

**Comment by timrae** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 1
           

Complexity increasing per file
==============================
- AnkiDroid/src/main/java/com/ichi2/utils/WebViewDebugging.java  2
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/timrae/Anki-Android/pullRequest?prid=5341529&bid=17487717)

File: AnkiDroid/src/main/java/com/ichi2/utils/WebViewDebugging.java (hunk 1 above)
Comment (timrae) [inline AnkiDroid/src/main/java/com/ichi2/utils/WebViewDebugging.java:7]: Codacy Issue found: [The utility class name 'WebViewDebugging' doesn't match '[A-Z][a-zA-Z0-9]+(Utils?|Helper)'](https://app.codacy.com/manual/timrae/Anki-Android/pullRequest?prid=5341529)

File: AnkiDroid/src/main/java/com/ichi2/anki/AnkiDroidApp.java (hunk 4 above)
Comment (mikehardy) [reply AnkiDroid/src/main/java/com/ichi2/anki/AnkiDroidApp.java]: I dunno - I just searched preferences in Android Studio for 'import' and didn't see anything that would turn on or off auto-organization. I kind of like them organized myself though it is diff noise. At least it's noise in a place that no one really looks at anyway (w.r.t. diffs) so long as they aren't unused (which codacy catches, so I just skim that chunk). meh

**Comment by mikehardy** [comment]
This is released as 2.10alpha66 as soon as google is done processing it

**Comment by david-allison-1** [comment]
Already in my local master. Thanks @mikehardy!

**Comment by david-allison-1** [comment]
@mikehardy This has broken `androidTest` on my device, but hasn't triggered any issues with the CI

```

java.lang.RuntimeException: Toggling of Web Contents Debugging must be done on the UI thread
at uk.a(PG:7)
at com.android.webview.chromium.WebViewChromiumFactoryProvider$1.setWebContentsDebuggingEnabled(PG:3)
at android.webkit.WebView.setWebContentsDebuggingEnabled(WebView.java:2138)
at com.ichi2.utils.WebViewDebugging.initializeDebugging(WebViewDebugging.java:14)
at com.ichi2.anki.AnkiDroidApp.onCreate(AnkiDroidApp.java:231)
at com.ichi2.anki.tests.ACRATest.setUp(ACRATest.java:52)
at java.lang.reflect.Method.invoke(Native Method)
at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:50)
at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:47)
at androidx.test.internal.runner.junit4.statement.RunBefores.evaluate(RunBefores.java:76)
at androidx.test.rule.GrantPermissionRule$RequestPermissionStatement.evaluate(GrantPermissionRule.java:134)
at org.junit.rules.RunRules.evaluate(RunRules.java:20)
at org.junit.runners.ParentRunner.runLeaf(ParentRunner.java:325)
at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:78)
at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:57)
at org.junit.runners.ParentRunner$3.run(ParentRunner.java:290)
at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:71)
at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:288)
at org.junit.runners.ParentRunner.access$000(ParentRunner.java:58)
at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:268)
at org.junit.runners.ParentRunner.run(ParentRunner.java:363)
at androidx.test.ext.junit.runners.AndroidJUnit4.run(AndroidJUnit4.java:104)
at org.junit.runners.Suite.runChild(Suite.java:128)
at org.junit.runners.Suite.runChild(Suite.java:27)
at org.junit.runners.ParentRunner$3.run(ParentRunner.java:290)
at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:71)
at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:288)
at org.junit.runners.ParentRunner.access$000(ParentRunner.java:58)
at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:268)
at org.junit.runners.ParentRunner.run(ParentRunner.java:363)
at org.junit.runner.JUnitCore.run(JUnitCore.java:137)
at org.junit.runner.JUnitCore.run(JUnitCore.java:115)
at androidx.test.internal.runner.TestExecutor.execute(TestExecutor.java:56)
at androidx.test.runner.AndroidJUnitRunner.onStart(AndroidJUnitRunner.java:392)
at android.app.Instrumentation$InstrumentationThread.run(Instrumentation.java:2252)
```

**Comment by david-allison-1** [comment]
Just to note: I'm fixing this, so no need for a revert.

**Comment by mikehardy** [comment]
I've struggled with this issue - there are some obvious explosions that CI doesn't catch. Back when I had a "test harness derby" and settled on robolectric I did so because espresso UI testing was unreliable.

But expressing "espresso UI testing was unreliable" needs to be a little more subtle, I found it unreliable in that attempting to drive the UI that way was prone to spurious failure because the elements moved or there were timing issues.

But I bet this (as well as a major boo-boo I did once that made the Reviewer crash during Activity.onCreate) might be caught with an espresso test that did nothing more than take a list of our Activities and load them. And that should be super reliable as the goal isn't interaction it is just making sure they start

**Comment by david-allison-1** [comment]
> I've struggled with this issue - there are some obvious explosions that CI doesn't catch. Back when I had a "test harness derby" and settled on robolectric I did so because espresso UI testing was unreliable.
> 
> But expressing "espresso UI testing was unreliable" needs to be a little more subtle, I found it unreliable in that attempting to drive the UI that way was prone to spurious failure because the elements moved or there were timing issues.
> 
> But I bet this (as well as a major boo-boo I did once that made the Reviewer crash during Activity.onCreate) might be caught with an espresso test that did nothing more than take a list of our Activities and load them. And that should be super reliable as the goal isn't interaction it is just making sure they start

From my understanding, it's just a testing issue. It works fine on my device, but the tests weren't running on the UI thread.

I'll take https://stackoverflow.com/a/30390340 as proof that this wasn't a bad change.

> The system does not create a separate thread for each instance of a component. All components that run in the same process are instantiated in the UI thread, and system calls to each component are dispatched from that thread. Consequently, methods that respond to system callbacks (such as onKeyDown() to report user actions or a lifecycle callback method) always run in the UI thread of the process.

**Comment by mikehardy** [comment]
Here is a UI test that tries to do a lot but if trimmed down might be useful / reliable https://github.com/ankidroid/Anki-Android/pull/5147/files#diff-61533c44c23f50b056c0ef73e0eb4341

And of course that whole suggestion might be off target but it's an idea ("start each activity in CI") on my back burner at least

**Comment by mikehardy** [comment]
out as 2.10alpha68 as soon as google is done processing it...

**Pull request merged**